numba/types/abstract.py

```diff
@@ -209,9 +249,213 @@ class Hashable(Type):
     """
+    @classmethod
+    def interface_check(cls, instance):
+        return cls.virtual_subclass_check(instance, [UserHashable])
+
+
+class UserHashable(Type):
+    """
+    For user-defined types that may define __hash__.
+    """
+    def is_hashable(self):
+        """
+        Returns True if the type is hashable.
+        `isinstance(self, Hasable)` will be True.
+        Must be overriden in subclass
+        """
+        raise NotImplementedError
+
+    def get_user_hash(self, context, sig):
+        """
+        Returns (call, callsig) where
+            * `res = call(builder, args)` emits the hashing operation to
+               `builder` with `args` and returns to `res`;
+            * `callsig` is the signature of `call`.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def interface_check(cls, instance):
+        return issubclass(type(instance), cls) and instance.is_hashable()
+
+
+class Ne(Type):
+    """
+    Base class for Ne types.
+    Subclass must provide ``__ne__``    
+    """
+    @classmethod
+    def interface_check(cls, instance):
+        if IS_PY3:
+            return cls.virtual_subclass_check(instance, [UserNe, UserEq])
+        else:
+            return cls.virtual_subclass_check(instance, [UserNe])
+
+
+class Eq(Type):
+    """
+    Base class for Eq types.
+    Subclass must provide ``__eq__``
+    In Py3, provides ``__ne__`` via inversion of ``__eq__``.
+    
+    If __eq__ is implemented and not __ne__, python semantic provides a 
+    default __ne__ implementation that uses __eq__.
+    If __ne__ is implemented and not __eq__, __eq__ fallbacks to `is`.    
+    """
+    @classmethod
+    def interface_check(cls, instance):
+        return cls.virtual_subclass_check(instance, [UserEq])
+
+
+class UserEq(Type):
+    """
+    For user-defined types that may define ``__eq__`` and ``__ne__``.
+    Virtually subclass Eq if `self.supports_eq()` returns True.
+    """
+    def supports_eq(self):
+        raise NotImplementedError
+
+    def get_user_eq(self, context, sig):
+        raise NotImplementedError
+
+    @classmethod
+    def interface_check(cls, instance):
+        return issubclass(type(instance), cls) and instance.supports_eq()
+        
+
+class UserNe(Type):
+    """
+    For user-defined types that may provide ``__ne__``.
+    Note that ``__ne__`` may be provided indirectly via ``__eq__``.
+    Virtually subclass Ne if `self.supports_ne()` returns True.
+    """
+
+    def supports_ne(self):
+        raise NotImplementedError
+
+    def get_user_ne(self, context, sig):
+        raise NotImplementedError
+
+    @classmethod
+    def interface_check(cls, instance):
+        typ = type(instance)
+        if IS_PY3:
+            return ((issubclass(typ, UserNe) and instance.supports_ne()) or
+                    (issubclass(typ, UserEq) and instance.supports_eq()))
+        else:
+            return issubclass(typ, UserNe) and instance.supports_ne()
+
+
+class Lt(Type):
```

Comment:
The distinct classes are needed because there are no default implementation for ordered comparisons and user-defined classes is allowed define some but not all of the comparisons.

They also help spelling the reflection rule for the comparison operators: i.e.  `<` is available if LHS type is a Lt or RHS type is a Gt.